e2e/prevent_delete_test.go

```diff
@@ -0,0 +1,291 @@
+		if cluster.Status.CurrentPrimaryIndex != 1 {
+			kubectlSafe(nil, "moco", "-n", "prevent-delete", "switchover", "test")
+			time.Sleep(10 * time.Second)
+		}
+		return cluster.Status.CurrentPrimaryIndex
+	}).Should(Equal(1))
+
+	kubectlSafe(nil, "moco", "mysql", "-n", "prevent-delete", "-u", "moco-writable", "test", "--",
+		"-e", "CREATE DATABASE test")
+	kubectlSafe(nil, "moco", "mysql", "-n", "prevent-delete", "-u", "moco-writable", "test", "--",
+		"-e", "CREATE TABLE test.t (i INT)")
+}
+
+func cleanupMySQL() {
+	cluster, err := getCluster("prevent-delete", "test")
+	Expect(err).NotTo(HaveOccurred())
+	primary := cluster.Status.CurrentPrimaryIndex
+	for i := 0; i < 3; i++ {
+		if i == primary {
+			continue
+		}
+		setSourceDelay(i, 0)
+	}
+	time.Sleep(10 * time.Second)
+
+	kubectlSafe(nil, "delete", "mysqlclusters", "-n", "prevent-delete", "--all")
+	verifyAllPodsDeleted("prevent-delete")
+}
+
+func setSourceDelay(index, delay int) {
+	kubectlSafe(nil, "moco", "mysql", "-n", "prevent-delete", "-u", "moco-admin", "--index", strconv.Itoa(index), "test", "--", "-e", "STOP REPLICA SQL_THREAD")
+	kubectlSafe(nil, "moco", "mysql", "-n", "prevent-delete", "-u", "moco-admin", "--index", strconv.Itoa(index), "test", "--", "-e", fmt.Sprintf("CHANGE REPLICATION SOURCE TO SOURCE_DELAY=%d", delay))
+	kubectlSafe(nil, "moco", "mysql", "-n", "prevent-delete", "-u", "moco-admin", "--index", strconv.Itoa(index), "test", "--", "-e", "START REPLICA")
+	if delay != 0 {
+		kubectlSafe(nil, "moco", "mysql", "-n", "prevent-delete", "-u", "moco-writable", "test", "--", "-e", "INSERT INTO test.t VALUES (1); COMMIT;")
+	}
+}
+
+var _ = Context("PreventDelete", func() {
+	if doUpgrade {
+		return
+	}
+
+	BeforeEach(func() {
+		initializeMySQL()
+	})
+
+	AfterEach(func() {
+		cleanupMySQL()
+	})
+
+	It("should add or remove prevent-delete annotation by replication delay", func() {
+		cluster, err := getCluster("prevent-delete", "test")
+		Expect(err).NotTo(HaveOccurred())
+		primary := cluster.Status.CurrentPrimaryIndex
+
+		// add prevent-delete annotation and wait for it to be removed
+		for i := 0; i < 3; i++ {
+			kubectlSafe(nil, "annotate", "pod", "-n", "prevent-delete", fmt.Sprintf("moco-test-%d", i), "moco.cybozu.com/prevent-delete=true")
+			Eventually(func() error {
+				out, err := kubectl(nil, "get", "pod", "-n", "prevent-delete", fmt.Sprintf("moco-test-%d", i), "-o", "json")
+				Expect(err).NotTo(HaveOccurred())
+				pod := &corev1.Pod{}
+				err = json.Unmarshal(out, pod)
+				Expect(err).NotTo(HaveOccurred())
+				if _, exists := pod.Annotations[constants.AnnPreventDelete]; exists {
+					return errors.New("annotation is not removed")
+				}
+				return nil
+			}).Should(Succeed())
+		}
+
+		// set huge replication delay
+		setSourceDelay(0, 10000)
+
+		// wait for prevent-delete annotation to be added
+		Eventually(func() error {
+			out, err := kubectl(nil, "get", "pod", "-n", "prevent-delete", fmt.Sprintf("moco-test-%d", primary), "-o", "json")
```

Comment:
```suggestion
			out, err := kubectl(nil, "get", "pod", "-n", "prevent-delete", cluster.PodName(primary), "-o", "json")
```

Similar statements are found elsewhere

Reply:
[271ce8b](https://github.com/cybozu-go/moco/pull/751/commits/271ce8b726b1aadaeb7396488c97041caefec1e2)